### ISSUE TITLE
[Perf] Changes at 3/21/2021 5:03:52 PM

### ISSUE THREAD

**Issue by DrewScoggins** [dotnet/runtime#50737]
### Run Information

Architecture | x64
-- | --
OS | Windows 10.0.18362
Baseline | [3961d91812c997af18ff1ebbdac53bcc4d84fb02](https://github.com/dotnet/runtime/commit/3961d91812c997af18ff1ebbdac53bcc4d84fb02)
Compare | [c07e1cf4a874d4f673f4ce0e15f384917e3e2b32](https://github.com/dotnet/runtime/commit/c07e1cf4a874d4f673f4ce0e15f384917e3e2b32)

### Regressions in Microsoft.Extensions.Primitives.StringSegmentBenchmark

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[TrimEnd](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_Windows 10.0.18362/Microsoft.Extensions.Primitives.StringSegmentBenchmark.TrimEnd.html>) | 13.04 ns | 14.98 ns | 1.15 | | | | [Trace](https://helixri107v0xdeko0k025g8.blob.core.windows.net/results-c7ffa8f36c8548389a/Collect%20Microsoft.Extensions.Primitives.StringSegmentBenchmark.TrimEnd/artifacts/BenchmarkDotNet.Artifacts/Microsoft.Extensions.Primitives.StringSegmentBenchmark.TrimEnd-20210329-235629.etl?sv=2019-07-07&se=2021-06-28T06%3A54%3A56Z&sr=c&sp=rl&sig=5R%2B0TX4hxJDpMjcfFgao0Q4m9XuXmKINlJ2gvd78%2Fmc%3D) | [Trace](https://helixri107v0xdeko0k025g8.blob.core.windows.net/results-411c87de6d5f4c718f/Collect%20Microsoft.Extensions.Primitives.StringSegmentBenchmark.TrimEnd/artifacts/BenchmarkDotNet.Artifacts/Microsoft.Extensions.Primitives.StringSegmentBenchmark.TrimEnd-20210329-235608.etl?sv=2019-07-07&se=2021-06-28T06%3A54%3A59Z&sr=c&sp=rl&sig=%2BJTdxVBgvS40uYMybjfdfNiLIpobmoUB%2FWX008RxMVA%3D)
[GetSegmentHashCode](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_Windows 10.0.18362/Microsoft.Extensions.Primitives.StringSegmentBenchmark.GetSegmentHashCode.html>) | 13.43 ns | 15.23 ns | 1.13 | | | | [Trace](https://helixri107v0xdeko0k025g8.blob.core.windows.net/results-a42cb462dfc74c1e94/Collect%20Microsoft.Extensions.Primitives.StringSegmentBenchmark.GetSegmentHashCode/artifacts/BenchmarkDotNet.Artifacts/Microsoft.Extensions.Primitives.StringSegmentBenchmark.GetSegmentHashCode-20210329-235752.etl?sv=2019-07-07&se=2021-06-28T06%3A55%3A03Z&sr=c&sp=rl&sig=SNRhWTSujvNtdkyKbD0N8Lresa2P5q0926yPq5%2B1UeA%3D) | [Trace](https://helixri107v0xdeko0k025g8.blob.core.windows.net/results-bb09dbfd5dcc4212b4/Collect%20Microsoft.Extensions.Primitives.StringSegmentBenchmark.GetSegmentHashCode/artifacts/BenchmarkDotNet.Artifacts/Microsoft.Extensions.Primitives.StringSegmentBenchmark.GetSegmentHashCode-20210329-235737.etl?sv=2019-07-07&se=2021-06-28T06%3A55%3A04Z&sr=c&sp=rl&sig=3nO%2BQyOJ4fkLHQf7fCr0%2BQS0tlsllIQn7uChPA4IpDk%3D)

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_30_2021/refs/heads/main_x64_Windows%2010.0.18362_Regression/Microsoft.Extensions.Primitives.StringSegmentBenchmark_1.png>)
![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_30_2021/refs/heads/main_x64_Windows%2010.0.18362_Regression/Microsoft.Extensions.Primitives.StringSegmentBenchmark_2.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_30_2021/refs/heads/main_x64_Windows%2010.0.18362_Regression/Microsoft.Extensions.Primitives.StringSegmentBenchmark.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Microsoft.Extensions.Primitives.StringSegmentBenchmark*'
```
<details>

### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-75aa3a3c-4975-452c-bbc3-470888aeedc868587cff37d436784/ccea48db-e751-4aa3-94e0-5658466dd473.zip?sv=2019-07-07&se=2021-04-20T08%3A37%3A29Z&sr=c&sp=rl&sig=zma87dXLZZAk5WwzTTss3etCRDGghPmQLDor1ArVE%2FQ%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-fa5ec340-4a62-498c-9582-3243549bd466267a47689aa49ada5/36cac4c4-0780-48e1-a54b-fab55d216301.zip?sv=2019-07-07&se=2021-04-20T21%3A16%3A29Z&sr=c&sp=rl&sig=mMmQjge8U5%2BjYENq0FU8OmY2%2FFrtz6a8knhp0FylfAU%3D>)

#### Microsoft.Extensions.Primitives.StringSegmentBenchmark.TrimEnd

```log

```
### Baseline Jit Disasm

```assembly
; Microsoft.Extensions.Primitives.StringSegmentBenchmark.TrimEnd()
       cmp       [rcx],ecx
       add       rcx,48
       jmp       near ptr Microsoft.Extensions.Primitives.StringSegment.TrimEnd()
; Total bytes of code 11
```
```assembly
; Microsoft.Extensions.Primitives.StringSegment.TrimEnd()
       push      r15
       push      r14
       push      rdi
       push      rsi
       push      rbp
       push      rbx
       sub       rsp,28
       mov       rsi,rcx
       mov       rdi,rdx
       mov       rcx,[rsi]
       mov       eax,[rsi+8]
       mov       edx,[rsi+0C]
       test      rcx,rcx
       jne       short M01_L00
       or        eax,edx
       jne       near ptr M01_L06
       xor       ebx,ebx
       xor       ebp,ebp
       jmp       short M01_L01
M01_L00:
       mov       r8d,eax
       mov       r9d,edx
       add       r8,r9
       mov       r9d,[rcx+8]
       cmp       r8,r9
       ja        near ptr M01_L06
       add       rcx,0C
       movsxd    rax,eax
       lea       rbx,[rcx+rax*2]
       mov       ebp,edx
M01_L01:
       lea       r14d,[rbp+0FFFF]
       test      r14d,r14d
       jl        short M01_L05
M01_L02:
       cmp       r14d,ebp
       jae       short M01_L07
       movsxd    rcx,r14d
       movzx     ecx,word ptr [rbx+rcx*2]
       cmp       ecx,100
       jae       short M01_L03
       cmp       ecx,100
       jae       short M01_L07
       movsxd    rcx,ecx
       mov       rax,7FFAC36708C0
       test      byte ptr [rcx+rax],80
       setne     r15b
       movzx     r15d,r15b
       test      r15d,r15d
       je        short M01_L05
       jmp       short M01_L04
M01_L03:
       call      System.Globalization.CharUnicodeInfo.GetIsWhiteSpace(Char)
       movzx     r15d,al
       test      r15d,r15d
       je        short M01_L05
M01_L04:
       dec       r14d
       test      r14d,r14d
       jge       short M01_L02
M01_L05:
       lea       r9d,[r14+1]
       mov       rcx,rsi
       mov       rdx,rdi
       xor       r8d,r8d
       call      Microsoft.Extensions.Primitives.StringSegment.Subsegment(Int32, Int32)
       mov       rax,rdi
       add       rsp,28
       pop       rbx
       pop       rbp
       pop       rsi
       pop       rdi
       pop       r14
       pop       r15
       ret
M01_L06:
       mov       ecx,21
       call      System.ThrowHelper.ThrowArgumentOutOfRangeException(System.ExceptionArgument)
       int       3
M01_L07:
       call      CORINFO_HELP_RNGCHKFAIL
       int       3
; Total bytes of code 223
```

### Compare Jit Disasm

```assembly
; Microsoft.Extensions.Primitives.StringSegmentBenchmark.TrimEnd()
       cmp       [rcx],ecx
       add       rcx,48
       jmp       near ptr Microsoft.Extensions.Primitives.StringSegment.TrimEnd()
; Total bytes of code 11
```
```assembly
; Microsoft.Extensions.Primitives.StringSegment.TrimEnd()
       push      r15
       push      r14
       push      rdi
       push      rsi
       push      rbp
       push      rbx
       sub       rsp,28
       mov       rsi,rcx
       mov       rdi,rdx
       mov       rcx,[rsi]
       mov       eax,[rsi+8]
       mov       edx,[rsi+0C]
       test      rcx,rcx
       jne       short M01_L00
       or        eax,edx
       jne       near ptr M01_L06
       xor       ebx,ebx
       xor       ebp,ebp
       jmp       short M01_L01
M01_L00:
       mov       r8d,eax
       mov       r9d,edx
       add       r8,r9
       mov       r9d,[rcx+8]
       cmp       r8,r9
       ja        near ptr M01_L06
       add       rcx,0C
       movsxd    rax,eax
       lea       rbx,[rcx+rax*2]
       mov       ebp,edx
M01_L01:
       lea       r14d,[rbp+0FFFF]
       test      r14d,r14d
       jl        short M01_L05
M01_L02:
       cmp       r14d,ebp
       jae       short M01_L07
       movsxd    rcx,r14d
       movzx     ecx,word ptr [rbx+rcx*2]
       cmp       ecx,100
       jae       short M01_L03
       cmp       ecx,100
       jae       short M01_L07
       movsxd    rcx,ecx
       mov       rax,7FFA5DB30B50
       test      byte ptr [rcx+rax],80
       setne     r15b
       movzx     r15d,r15b
       test      r15d,r15d
       je        short M01_L05
       jmp       short M01_L04
M01_L03:
       call      System.Globalization.CharUnicodeInfo.GetIsWhiteSpace(Char)
       movzx     r15d,al
       test      r15d,r15d
       je        short M01_L05
M01_L04:
       dec       r14d
       test      r14d,r14d
       jge       short M01_L02
M01_L05:
       lea       r9d,[r14+1]
       mov       rcx,rsi
       mov       rdx,rdi
       xor       r8d,r8d
       call      Microsoft.Extensions.Primitives.StringSegment.Subsegment(Int32, Int32)
       mov       rax,rdi
       add       rsp,28
       pop       rbx
       pop       rbp
       pop       rsi
       pop       rdi
       pop       r14
       pop       r15
       ret
M01_L06:
       mov       ecx,21
       call      System.ThrowHelper.ThrowArgumentOutOfRangeException(System.ExceptionArgument)
       int       3
M01_L07:
       call      CORINFO_HELP_RNGCHKFAIL
       int       3
; Total bytes of code 223
```

#### Microsoft.Extensions.Primitives.StringSegmentBenchmark.GetSegmentHashCode

```log

```
### Baseline Jit Disasm

```assembly
; Microsoft.Extensions.Primitives.StringSegmentBenchmark.GetSegmentHashCode()
       push      rdi
       push      rsi
       sub       rsp,28
       add       rcx,18
       mov       rdx,[rcx]
       mov       esi,[rcx+8]
       mov       edi,[rcx+0C]
       test      rdx,rdx
       je        short M00_L01
       mov       ecx,esi
       mov       r8d,edi
       add       rcx,r8
       mov       r8d,[rdx+8]
       cmp       rcx,r8
       ja        short M00_L02
       add       rdx,0C
       movsxd    rcx,esi
       lea       rcx,[rdx+rcx*2]
       mov       edx,edi
M00_L00:
       add       edx,edx
       mov       r8d,0B01B7655
       mov       r9d,1E7BB2D5
       call      System.Marvin.ComputeHash32(Byte ByRef, UInt32, UInt32, UInt32)
       nop
       add       rsp,28
       pop       rsi
       pop       rdi
       ret
M00_L01:
       or        esi,edi
       jne       short M00_L02
       xor       ecx,ecx
       xor       edx,edx
       jmp       short M00_L00
M00_L02:
       mov       ecx,21
       call      System.ThrowHelper.ThrowArgumentOutOfRangeException(System.ExceptionArgument)
       int       3
; Total bytes of code 102
```
```assembly
; System.Marvin.ComputeHash32(Byte ByRef, UInt32, UInt32, UInt32)
       cmp       edx,8
       jb        near ptr M01_L04
       mov       eax,edx
       shr       eax,3
M01_L00:
       add       r8d,[rcx]
       mov       r10d,[rcx+4]
       xor       r9d,r8d
       rol       r8d,14
       add       r8d,r9d
       rol       r9d,9
       xor       r9d,r8d
       rol       r8d,1B
       add       r8d,r9d
       rol       r9d,13
       add       r8d,r10d
       xor       r9d,r8d
       rol       r8d,14
       add       r8d,r9d
       rol       r9d,9
       xor       r9d,r8d
       rol       r8d,1B
       add       r8d,r9d
       rol       r9d,13
       add       rcx,8
       dec       eax
       jne       short M01_L00
       test      dl,4
       je        short M01_L02
M01_L01:
       add       r8d,[rcx]
       xor       r9d,r8d
       rol       r8d,14
       add       r8d,r9d
       rol       r9d,9
       xor       r9d,r8d
       rol       r8d,1B
       add       r8d,r9d
       rol       r9d,13
M01_L02:
       mov       eax,edx
       and       rax,7
       mov       r10d,[rcx+rax+0FFFC]
       not       edx
       shl       edx,3
       shr       r10d,8
       or        r10d,80000000
       mov       ecx,edx
       shr       r10d,cl
M01_L03:
       add       r8d,r10d
       mov       eax,r9d
       xor       eax,r8d
       mov       edx,r8d
       rol       edx,14
       add       edx,eax
       rol       eax,9
       xor       eax,edx
       rol       edx,1B
       add       edx,eax
       rol       eax,13
       xor       eax,edx
       mov       r8d,edx
       rol       r8d,14
       add       r8d,eax
       rol       eax,9
       xor       eax,r8d
       rol       r8d,1B
       add       r8d,eax
       rol       eax,13
       xor       eax,r8d
       ret
M01_L04:
       cmp       edx,4
       jae       near ptr M01_L01
       mov       r10d,80
       test      dl,1
       jne       short M01_L06
M01_L05:
       test      dl,2
       je        short M01_L03
       shl       r10d,10
       movzx     edx,word ptr [rcx]
       or        r10d,edx
       jmp       short M01_L03
M01_L06:
       mov       r10d,edx
       and       r10,2
       movzx     r10d,byte ptr [rcx+r10]
       or        r10d,8000
       jmp       short M01_L05
; Total bytes of code 276
```

### Compare Jit Disasm

```assembly
; Microsoft.Extensions.Primitives.StringSegmentBenchmark.GetSegmentHashCode()
       push      rdi
       push      rsi
       sub       rsp,28
       add       rcx,18
       mov       rdx,[rcx]
       mov       esi,[rcx+8]
       mov       edi,[rcx+0C]
       test      rdx,rdx
       je        short M00_L01
       mov       ecx,esi
       mov       r8d,edi
       add       rcx,r8
       mov       r8d,[rdx+8]
       cmp       rcx,r8
       ja        short M00_L02
       add       rdx,0C
       movsxd    rcx,esi
       lea       rcx,[rdx+rcx*2]
       mov       edx,edi
M00_L00:
       add       edx,edx
       mov       r8d,684B9E2E
       mov       r9d,0E01CA80B
       call      System.Marvin.ComputeHash32(Byte ByRef, UInt32, UInt32, UInt32)
       nop
       add       rsp,28
       pop       rsi
       pop       rdi
       ret
M00_L01:
       or        esi,edi
       jne       short M00_L02
       xor       ecx,ecx
       xor       edx,edx
       jmp       short M00_L00
M00_L02:
       mov       ecx,21
       call      System.ThrowHelper.ThrowArgumentOutOfRangeException(System.ExceptionArgument)
       int       3
; Total bytes of code 102
```
```assembly
; System.Marvin.ComputeHash32(Byte ByRef, UInt32, UInt32, UInt32)
       cmp       edx,8
       jb        near ptr M01_L04
       mov       eax,edx
       shr       eax,3
M01_L00:
       add       r8d,[rcx]
       mov       r10d,[rcx+4]
       xor       r9d,r8d
       rol       r8d,14
       add       r8d,r9d
       rol       r9d,9
       xor       r9d,r8d
       rol       r8d,1B
       add       r8d,r9d
       rol       r9d,13
       add       r8d,r10d
       xor       r9d,r8d
       rol       r8d,14
       add       r8d,r9d
       rol       r9d,9
       xor       r9d,r8d
       rol       r8d,1B
       add       r8d,r9d
       rol       r9d,13
       add       rcx,8
       dec       eax
       jne       short M01_L00
       test      dl,4
       je        short M01_L02
M01_L01:
       add       r8d,[rcx]
       xor       r9d,r8d
       rol       r8d,14
       add       r8d,r9d
       rol       r9d,9
       xor       r9d,r8d
       rol       r8d,1B
       add       r8d,r9d
       rol       r9d,13
M01_L02:
       mov       eax,edx
       and       rax,7
       mov       r10d,[rcx+rax+0FFFC]
       not       edx
       shl       edx,3
       shr       r10d,8
       or        r10d,80000000
       mov       ecx,edx
       shr       r10d,cl
M01_L03:
       add       r8d,r10d
       mov       eax,r9d
       xor       eax,r8d
       mov       edx,r8d
       rol       edx,14
       add       edx,eax
       rol       eax,9
       xor       eax,edx
       rol       edx,1B
       add       edx,eax
       rol       eax,13
       xor       eax,edx
       mov       r8d,edx
       rol       r8d,14
       add       r8d,eax
       rol       eax,9
       xor       eax,r8d
       rol       r8d,1B
       add       r8d,eax
       rol       eax,13
       xor       eax,r8d
       ret
M01_L04:
       cmp       edx,4
       jae       near ptr M01_L01
       mov       r10d,80
       test      dl,1
       jne       short M01_L06
M01_L05:
       test      dl,2
       je        short M01_L03
       shl       r10d,10
       movzx     edx,word ptr [rcx]
       or        r10d,edx
       jmp       short M01_L03
M01_L06:
       mov       r10d,edx
       and       r10,2
       movzx     r10d,byte ptr [rcx+r10]
       or        r10d,8000
       jmp       short M01_L05
; Total bytes of code 276
```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>
### Run Information

Architecture | x64
-- | --
OS | Windows 10.0.18362
Baseline | [3961d91812c997af18ff1ebbdac53bcc4d84fb02](https://github.com/dotnet/runtime/commit/3961d91812c997af18ff1ebbdac53bcc4d84fb02)
Compare | [c07e1cf4a874d4f673f4ce0e15f384917e3e2b32](https://github.com/dotnet/runtime/commit/c07e1cf4a874d4f673f4ce0e15f384917e3e2b32)

### Regressions in System.Collections.CopyTo&lt;Int32&gt;

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[Memory](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_Windows 10.0.18362/System.Collections.CopyTo(Int32).Memory(Size%3a%202048).html>) | 152.19 ns | 174.09 ns | 1.14 | | |
[Array](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_Windows 10.0.18362/System.Collections.CopyTo(Int32).Array(Size%3a%202048).html>) | 157.88 ns | 191.82 ns | 1.21 | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_30_2021/refs/heads/main_x64_Windows%2010.0.18362_Regression/System.Collections.CopyTo(Int32)_1.png>)
![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_30_2021/refs/heads/main_x64_Windows%2010.0.18362_Regression/System.Collections.CopyTo(Int32)_2.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/03_30_2021/refs/heads/main_x64_Windows%2010.0.18362_Regression/System.Collections.CopyTo(Int32).html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'System.Collections.CopyTo&lt;Int32&gt;*'
```
<details>

### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-75aa3a3c-4975-452c-bbc3-470888aeedc868587cff37d436784/ccea48db-e751-4aa3-94e0-5658466dd473.zip?sv=2019-07-07&se=2021-04-20T08%3A37%3A29Z&sr=c&sp=rl&sig=zma87dXLZZAk5WwzTTss3etCRDGghPmQLDor1ArVE%2FQ%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-fa5ec340-4a62-498c-9582-3243549bd466267a47689aa49ada5/36cac4c4-0780-48e1-a54b-fab55d216301.zip?sv=2019-07-07&se=2021-04-20T21%3A16%3A29Z&sr=c&sp=rl&sig=mMmQjge8U5%2BjYENq0FU8OmY2%2FFrtz6a8knhp0FylfAU%3D>)

#### System.Collections.CopyTo&lt;Int32&gt;.Memory(Size: 2048)

```log

```
#### System.Collections.CopyTo&lt;Int32&gt;.Array(Size: 2048)

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

**Comment by JulieLeeMSFT** [comment]
@kunalspathak PTAL.

**Comment by danmoseley** [comment]
@DrewScoggins would it be possible to update your script that creates these to add one more link -- a one click to the diff?

eg in this issue you have 
https://github.com/dotnet/runtime/commit/3961d91812c997af18ff1ebbdac53bcc4d84fb02 and https://github.com/dotnet/runtime/commit/c07e1cf4a874d4f673f4ce0e15f384917e3e2b32

those only show the endpoints, but if you constructed this url `https://github.com/dotnet/runtime/compare/3961d91812c997af18ff1ebbdac53bcc4d84fb02...c07e1cf4a874d4f673f4ce0e15f384917e3e2b32` the [view is much more convenient, showing all commits and the entire diff](https://github.com/dotnet/runtime/compare/3961d91812c997af18ff1ebbdac53bcc4d84fb02...c07e1cf4a874d4f673f4ce0e15f384917e3e2b32).

If the interval is largeish, this URL may not render, but much of the time it will and it does in this case. It immediately shows all the intervening commits.

**Comment by danmoseley** [comment]
In this case, it's hard to imagine any of the commits could have caused this. Is it possible there was some environmental change?

**Comment by DrewScoggins** [comment]
Yes, I have been meaning to add this back in now that we are able to generate tighter diffs, but have not gotten around to it. I will make sure this gets in for the triage tomorrow.

**Comment by DrewScoggins** [comment]
There was no change in the performance repo over this gap. During our triage meeting we thought it might be related to degraded PGO data, but we were not sure.

**Comment by kunalspathak** [comment]
@AndyAyersMS - do you want to look into this one as you might be able to quickly flag if it is related to stale PGO or not?

**Comment by AndyAyersMS** [comment]
Sure.

**Comment by kunalspathak** [comment]
I am assigning this to you @AndyAyersMS .

**Comment by AndyAyersMS** [comment]
Historical data for these tests. The first two seems like they're likely regressions from PGO.

![newplot (48)](https://user-images.githubusercontent.com/10121823/121604769-f9386c00-c9ff-11eb-9ef2-26a068e3112b.png)
![newplot (46)](https://user-images.githubusercontent.com/10121823/121604679-d312cc00-c9ff-11eb-9738-a173cd37c4ae.png)
![newplot (49)](https://user-images.githubusercontent.com/10121823/121604803-0ead9600-ca00-11eb-92a6-bbfa53422b79.png)
![newplot (50)](https://user-images.githubusercontent.com/10121823/121604834-1ff6a280-ca00-11eb-9548-377b784b2c25.png)

**Comment by AndyAyersMS** [comment]
Comparing StaticPGO (blue) vs NoPgo (purple) `TrimEnd` does not look like a PGO regression, since the new behavior persists even without PGO data. 

![newplot (51)](https://user-images.githubusercontent.com/10121823/121605586-85975e80-ca01-11eb-8f55-28f92c1d3321.png)

Best guess for regression is #47133 or perhaps #45022 based on the first and second bounce up here in the detail view:

![newplot (54)](https://user-images.githubusercontent.com/10121823/121606331-ccd21f00-ca02-11eb-9a64-91e4d68d65b9.png)



However `GetHashCode` looks like it may well be a PGO regression.

![newplot (52)](https://user-images.githubusercontent.com/10121823/121605875-fcccf280-ca01-11eb-9f35-10938dd452c5.png)

**Comment by danmoseley** [comment]
cc @GrabYourPitchforks for #45022

**Comment by GrabYourPitchforks** [comment]
At first glance, I'd say it's a combination of: (a) we're now detecting struct tearing before the loop is entered; and (b) there may be duplicate bounds checks within the loop.

We can't do anything about (a) since it fixed an existing bug in `StringSegment`, but we can probably fudge the loop code to address (b). Should I investigate and send a PR?

**Comment by AndyAyersMS** [comment]
> Should I investigate and send a PR?

Let me look at the codegen. I'll post what I find here.

**Comment by AndyAyersMS** [comment]
Here's the current inner loop for `StringSegment:TrimEnd():StringSegment:this`
```asm
G_M47402_IG06:              ;; offset=0059H
       443BF5               cmp      r14d, ebp
       737A                 jae      SHORT G_M47402_IG13
       4963CE               movsxd   rcx, r14d
       0FB70C4B             movzx    rcx, word  ptr [rbx+2*rcx]
       81F900010000         cmp      ecx, 256
       7328                 jae      SHORT G_M47402_IG08
						;; bbWeight=4    PerfScore 19.00
G_M47402_IG07:              ;; offset=006DH
       81F900010000         cmp      ecx, 256
       7363                 jae      SHORT G_M47402_IG13
       4863C9               movsxd   rcx, ecx
       48B87800E272FE7F0000 mov      rax, 0x7FFE72E20078
       F6040180             test     byte  ptr [rcx+rax], 128
       410F95C7             setne    r15b
       450FB6FF             movzx    r15, r15b
       4585FF               test     r15d, r15d
       7418                 je       SHORT G_M47402_IG10
       EB0E                 jmp      SHORT G_M47402_IG09
						;; bbWeight=2    PerfScore 16.50
G_M47402_IG08:              ;; offset=0095H
       E85EF5FFFF           call     CharUnicodeInfo:GetIsWhiteSpace(ushort):bool
       440FB6F8             movzx    r15, al
       4585FF               test     r15d, r15d
       7408                 je       SHORT G_M47402_IG10
						;; bbWeight=2    PerfScore 5.00
G_M47402_IG09:              ;; offset=00A3H
       41FFCE               dec      r14d
       4585F6               test     r14d, r14d
       7DAE                 jge      SHORT G_M47402_IG06
```
`IG07` has a redundant test and an ugly bool materialization.

**Comment by AndyAyersMS** [comment]
For `GetSegmentHashCode` both the benchmark method and the workhorse callee `Marvin:ComputeHash32` are impacted by PGO.  I can repro a regression locally (9.4 with Static PGO, 8.9 without). I can't selectively disable PGO per method just yet; perhaps that would be a nice addition.

In `GetSegmentHashCode` the empty span construction case is seen as unlikely and reordered; unfortunately this leads to a few spills and so higher prolog/epilog costs. 

In `Marvin:ComputeHash32` the code is reordered to favor strings 8 bytes or longer; here the test string is 12 bytes and so the reordering actually should slightly favor this test case. 

Perfview analysis not yet conclusive... 

![image](https://user-images.githubusercontent.com/10121823/121622871-8f31be00-ca23-11eb-9b11-cfc2366516cc.png)

**Comment by GrabYourPitchforks** [comment]
@AndyAyersMS I can rewrite the inner loop as follows to elide the bounds check due to the reverse iterator.

```cs
ReadOnlySpan<char> span = AsSpan();
int i = span.Length - 1;
while (true)
{
    if ((uint)i >= (uint)span.Length || !char.IsWhiteSpace(span[i])) { break; }
    i--;
}
return Subsegment(0, i + 1);
```

For some reason, if I use the pattern `for (int i = span.Length - 1; (uint)i < (uint)span.Length; i--)`, the bounds check does not get elided. Putting the comparison within the `{ /* ... */ }` brackets is the only way I've found to elide the check.

For `char.IsWhiteSpace`, I haven't found a good way to avoid the duplicate bounds check against _Latin1CharInfo.Length_. But I can use the `return (condition) ? true : false;` pattern on line 273 below to elide the unnecessary `setcc` / `movzx` / `test` bounce.

https://github.com/dotnet/runtime/blob/ddbce77f82aa22d66905ec017d01dbdb8dd0ca1d/src/libraries/System.Private.CoreLib/src/System/Char.cs#L269-L277

**Comment by AndyAyersMS** [comment]
> the bounds check does not get elided.

Bounds check elimination is very pattern-sensitive. I'd have to drill in to see why it's not able to handle the above. Could be any number of things that throw it off.

The new redundant-branch opt phase should in principle be able to remove the redundant array bounds check, but it currently doesn't handle array bounds checks, as the control flow they cause is implicit during optimization. I can look into this sometime down the road.

Do you have a sense of how much perf we get back with these changes?

**Comment by AndyAyersMS** [comment]
For `GetSegmentHashCode`, VTune seems to suggest the issue is indeed with the pgo codegen for `GetSegmentHashCode` itself, and not for `Marvin:ComputeHash32`.

It is a bit clunky to investigate BDN tests with VTune. I run under BDN initially with `--keepFiles` and fix the number of iterations, then find the generated test job file under artifacts (Job-NNN.notcs) and modify this not to require host acknowledgments, rebuild the project, and run it under a release corerun with checked jit (so I can get jit disasm). Then I guestimate where in the process it started doing the real measurements (say a one-second slice close to process end) and compare the VTune profiles for the different configs.

Doing that for the above shows that for PGO we spend relatively more time in `GetSegmentHashCode` because it is less effective in making progress.
| Config | Method |  CPI | % of Interval | 
| --- | --- | ---- | --- |
| No PGO | ComputeHash32 | 0.348 | 77.3 |
| PGO | ComputeHash32 | 0.338 | 75.4 |
| No PGO | GetSegmentHashCode | 0.277 | 15.4 |
| PGO | GetSegmentHashCode  | 0.362 | 16.9 |

I am going to modify `COMPlus_JitDisablePGO` to allow it to be specified on a per-method basis; if the above is correct then disabling PGO just for `GetSegmentHashCode` should regain the lost perf (and then some).

**Comment by kunalspathak** [comment]
> It is a bit clunky to investigate BDN tests with VTune. I run under BDN initially with --keepFiles and fix the number of iterations, then find the generated test job file under artifacts (Job-NNN.notcs) and modify this not to require host acknowledgments, rebuild the project, and run it under a release corerun with checked jit (so I can get jit disasm).

Is it possible to just extract the benchmark in Hello world console app and run with `COMPlus_TieredCompilation=0` or have a loop that calls benchmark if you want to have tiering happen? Most of the time, this approach worked for me and I could eliminate the BDN repro process.

**Comment by AndyAyersMS** [comment]
> I am going to modify COMPlus_JitDisablePGO to allow it to be specified on a per-method basis

I realized I already have (in checked jits) `COMPlus_JitEnablePGORange` that works by method hash, so didn't need to modify the above. Using this to control which method can use the Static PGO data, I get the following matrix of results:

| ComputeHash32 | GetSegmentHashCode | Result |
| ---- | ---- | ---- |
| No PGO | No PGO | 8.88 |
| No PGO | PGO | 9.82 |
| PGO | No PGO | 8.83 |
| PGO | PGO | 9.48 |

This confirms the regression is coming from the application of Static PGO data to `GetSegmentHashCode`. Here's the diff in the codegen:

![image](https://user-images.githubusercontent.com/10121823/121739900-22650500-cab1-11eb-8121-f42d725d33ac.png)

Basically PGO identifies a block as rarely run; this block gets moved to the end of the method, and this rearrangement alters the allocation patterns so that there are now callee saves registers, which increases the per-call costs, regressing the hot path which calls the hash method.

The callee saves are for values live across a call (in the "no pgo" version these lifetimes aren't call crossing) but in this case the post-call uses are cold, and so I wonder if we'd be better off not using callee saves here (we still might need to spill so perhaps this doesn't accomplish much, though if we could realize that the backing field is not modified perhaps we could avoid spilling too).

@kunalspathak looks like this is something for LSRA to think about.

**Comment by AndyAyersMS** [comment]
> Is it possible to just extract the benchmark in Hello world console app

Could be -- that probably would have worked here. But I was also curious to try running BDN under VTune to see how feasible/useful it was... turns out it is a mixed bag.

**Comment by AndyAyersMS** [comment]
Back over to you, @kunalspathak

**Comment by kunalspathak** [comment]
Thanks for the analysis @AndyAyersMS . From what I understand, there is 1 regression in `GetSegmentHashCode()` that is due to register allocation. I believe for `TrimEnd()`, @GrabYourPitchforks  is making some changes? `CopyTo<Int32>.*` benchmarks can be ignored because we regained the performance.

**Comment by AndyAyersMS** [comment]
Right.

**Comment by kunalspathak** [comment]
I will have to look into the register allocation issue post .NET 6.

**Comment by kunalspathak** [comment]
The performance has improved overall since the issue was first opened.

<img width="727" alt="image" src="https://user-images.githubusercontent.com/12488060/209016677-c632cf42-f848-4a01-8409-ff428ed01c40.png">


There has been separate experiments conducted to tweak the `calleeSave` register preferences https://github.com/dotnet/runtime/pull/76904 . There was some TP cost to it, but we can re-evaluate it. There is nothing actionable for this issue to keep open.
